### PR TITLE
Update Extension and Support Klarna Payments

### PR DESCRIPTION
**Pull request by knorrli**
This MR converts the gem into a solidus extension conforming to the solidus extension standards according to [solidus_dev_support](https://github.com/solidusio/solidus_dev_support).
In addition, `solidus_six_saferpay` now supports Solidus 3, most importantly the switch from `first-` and `lastname` to a single name attribute for addresses.

## Notable Changes

- **Support for Solidus 3 by extracting first- and lastname from the single `Address#name` attribute. By default this depends on the behaviour of solidus' `Spree::Address::Name` which assumes that the first word of a name is the first name and everything else is the last name.**
- Configurable class to customize name extraction from addresses according to business needs
-  Added line item information to information sent to SIX when initializing payment
- Configurable class to provide params for initializing payment. This allows host applications to easily customize payment initialize params

### CODE REVIEW

File: .github/stale.yml [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
   - pinned
   - security
 # Label to use when marking an issue as stale
-staleLabel: wontfix
+staleLabel: stale
 # Comment to post when marking an issue as stale. Set to `false` to disable
 markComment: >
   This issue has been automatically marked as stale because it has not had
-  recent activity. It will be closed if no further activity occurs. Thank you
+  recent activity. It might be closed if no further activity occurs. Thank you
```
Comment (Drinion) [inline .github/stale.yml:14]: Must closing an issue be done manually?

File: app/controllers/spree/solidus_six_saferpay/checkout_controller.rb [hunk 2]
```diff
@@ -74,6 +63,17 @@ def success
           return
         end
 
+        saferpay_payment = Spree::SixSaferpayPayment.where(order_id: @order.id).order(:created_at).last
```
Comment (Drinion) [inline app/controllers/spree/solidus_six_saferpay/checkout_controller.rb]: Consider implementing a method in `SixSaferpayPayment` to return the last (current?) payment.

File: app/controllers/spree/solidus_six_saferpay/checkout_controller.rb [hunk 3]
```diff
@@ -139,20 +138,24 @@ def fail
 
       private
 
-      def initialize_payment(order, payment_method)
-        raise NotImplementedError, "Must be implemented in PaymentPageCheckoutController or TransactionCheckoutController"
+      def initialize_payment(_order, _payment_method)
```
Comment (Drinion) [inline app/controllers/spree/solidus_six_saferpay/checkout_controller.rb:141]: Why pass arguments, if they're not being used?

File: app/controllers/spree/solidus_six_saferpay/transaction/checkout_controller.rb [hunk 4]
```diff
@@ -1,9 +1,10 @@
+# frozen_string_literal: true
+
 module Spree
   module SolidusSixSaferpay
     module Transaction
-    # explicit parent must be stated, otherwise Spree::CheckoutController has precendence
+      # explicit parent must be stated, otherwise Spree::CheckoutController has precendence
```
Comment (Drinion) [inline app/controllers/spree/solidus_six_saferpay/transaction/checkout_controller.rb]: Change `precendence` to `precedence`.

File: app/models/spree/six_saferpay_payment.rb [hunk 5]
```diff
@@ -32,27 +34,21 @@ def address
 
     # This memoization only works if response_payment_means is not nil.
```
Comment (Drinion) [inline app/models/spree/six_saferpay_payment.rb:39]: Change `memoization` to `memorization`.

File: app/services/spree/solidus_six_saferpay/authorize_payment.rb [hunk 6]
```diff
@@ -26,22 +28,22 @@ def success?
...
       def saferpay_payment_attributes(saferpay_response)
         payment_means = saferpay_response.payment_means
-        brand = payment_means.brand
         card = payment_means.card
 
         attributes = {
           transaction_id: saferpay_response.transaction.id,
           transaction_status: saferpay_response.transaction.status,
-          transaction_date: DateTime.parse(saferpay_response.transaction.date),
+          transaction_date: DateTime.parse(saferpay_response.transaction.date).in_time_zone(Time.zone),
```
Comment (Drinion) [inline app/services/spree/solidus_six_saferpay/authorize_payment.rb]: Why is this necessary? Isn't this information already included `.date`?

File: app/services/spree/solidus_six_saferpay/initialize_payment.rb [hunk 7]
```diff
@@ -47,7 +51,7 @@ def saferpay_payment_attributes(api_response)
           order: order,
           payment_method: payment_method,
           token: api_response.token,
-          expiration: DateTime.parse(api_response.expiration),
+          expiration: DateTime.parse(api_response.expiration).in_time_zone(Time.zone),
```
Comment (Drinion) [inline app/services/spree/solidus_six_saferpay/initialize_payment.rb]: Same as before. Why is `.in_time_zone(Time.zone)` necessary?

File: lib/solidus_six_saferpay/address_name_extractor.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+module SolidusSixSaferpay
+  class AddressNameExtractor
+    class UnsafeNameExtractionError < StandardError
+    end
+
+    attr_reader :address_name
+
+    def initialize(address)
+      @address_name = extract_names(address)
```
Comment (Drinion) [inline lib/solidus_six_saferpay/address_name_extractor.rb]: IMO `extract_names(arg)` should be changed to `extract_name(arg)`.

File: lib/solidus_six_saferpay/gateway.rb [hunk 9]
```diff
@@ -52,12 +55,11 @@ def capture(_amount, transaction_id, options={})
         capture_response,
         { authorization: capture_response.capture_id }
       )
-
     rescue SixSaferpay::Error => e
       handle_error(e, capture_response)
     end
 
-    def void(transaction_id, options = {})
+    def void(transaction_id, _options = {})
```
Comment (Drinion) [inline lib/solidus_six_saferpay/gateway.rb:62]: Again, why pass unused arguments?

File: lib/solidus_six_saferpay/line_item_type_deductor.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  # This very simple default implementation can be overridden by configuring
+  # SolidusSixSaferpay::Config.line_item_type_deductor_class
+  class LineItemTypeDeductor
+    attr_reader :line_item
+
+    def initialize(line_item)
+      @line_item = line_item
+    end
+
+    # This very dumb default implementation is based on the fact that by
```
Comment (Drinion) [inline lib/solidus_six_saferpay/line_item_type_deductor.rb]: Maybe `This implementation is based on the fact that by default...` ;)

File: lib/solidus_six_saferpay/payment_initialize_params.rb [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        last_name: address_name.last_name,
+        date_of_birth: nil,
+        company: nil,
+        gender: nil,
+        legal_form: nil,
+        street: solidus_address.address1,
+        street2: solidus_address.address2,
+        zip: solidus_address.zipcode,
+        city: solidus_address.city,
+        country_subdevision_code: nil,
```
Comment (Drinion) [inline lib/solidus_six_saferpay/payment_initialize_params.rb:58]: Change `country_subdevision_code` to `country_subdivision_code`.

File: spec/models/spree/six_saferpay_payment_spec.rb [hunk 12]
```diff
@@ -14,14 +15,29 @@ module Spree
...
-        expect(Spree::Payment.count).to eq(0)
-        solidus_payment = payment.create_solidus_payment!
-        expect(Spree::Payment.count).to eq(1)
+      subject(:solidus_payment) { payment.create_solidus_payment! }
+
+      it 'creates a Solidus::Payment' do
+        expect { payment.create_solidus_payment! }.to change { Spree::Payment.count }.from(0).to(1)
+      end
+
+      it 'sets the order' do
```
Comment (Drinion) [inline spec/models/spree/six_saferpay_payment_spec.rb]: It sets the order of what?

File: spec/services/spree/solidus_six_saferpay/cancel_authorized_payment_spec.rb [hunk 13]
```diff
@@ -29,30 +28,36 @@ module SolidusSixSaferpay
           let(:payment) { create(:six_saferpay_payment) }
 
           it 'does not cancel the payment' do
-            expect(gateway).not_to receive(:void)
+            allow(gateway).to receive(:void)
```
Comment (Drinion) [inline spec/services/spree/solidus_six_saferpay/cancel_authorized_payment_spec.rb:31]: Why is `expect(gateway).not_to receive(:void)` changed to this (and 35) line? It seems to do the same thing.

File: spec/services/spree/solidus_six_saferpay/initialize_transaction_spec.rb [hunk 14]
```diff
@@ -3,34 +3,32 @@
...
             response_header: SixSaferpay::ResponseHeader.new(request_id: 'test', spec_version: 'test'),
-            token: token,
-            expiration: expiration,
-            liability_shift: nil, # this is empty because we don't submit PaymentMeans on initialize already
-            redirect_required: true, 
-            redirect: SixSaferpay::Redirect.new(redirect_url: redirect_url, payment_means_required: true) # since we don't provide PaymentMeans on initialize, we must always redirect
+            token: '234uhfh78234hlasdfh8234e1234',
+            expiration: '2015-01-30T12:45:22.258+01:00',
+
+            # this is empty because we don't submit PaymentMeans on initialize already
```
Comment (Drinion) [inline spec/services/spree/solidus_six_saferpay/initialize_transaction_spec.rb]: Change to `this is empty because PaymentMeans are not being submitted on initialize`.

File: spec/services/spree/solidus_six_saferpay/initialize_transaction_spec.rb [hunk 15]
```diff
@@ -3,34 +3,32 @@
...
+            liability_shift: nil,
+
+            # since we don't provide PaymentMeans on initialize, we must always redirect
```
Comment (Drinion) [inline spec/services/spree/solidus_six_saferpay/initialize_transaction_spec.rb]: Change to `it must always be redirected, since PaymentMeans are not provided on initialize`.

**Review by Drinion — COMMENTED** [review]
Thank you for this MR! I've only found some minor issues, such as spelling errors and some code that is unclear to me.

File: .github/stale.yml (hunk 1 above)
Comment (knorrli) [reply .github/stale.yml:14]: I don't know since I'm not familiar with github automation. This file (and some others) were autogenerated by `solidus_dev_support` to match the setup of standard solidus extensions.
I doubt that it will happen but we will find out if someone opens an issue (and we let it go stale)

File: app/controllers/spree/solidus_six_saferpay/checkout_controller.rb (hunk 3 above)
Comment (knorrli) [reply app/controllers/spree/solidus_six_saferpay/checkout_controller.rb:141]: I pass them here to document the interface of this method.
Ruby has no real way of doing `abstract` methods, so my preferred way of handling this is to implement the method in the base class to raise an error, and override this method in child classes with child-specific functionality.
This means that the arguments are technically useless in the base class method, but I still want to have them so that one can directly see the method signature and expected arguments when reading the code of the base class (because this is what you read when you are trying to figure out how to implement a child class).

File: app/controllers/spree/solidus_six_saferpay/transaction/checkout_controller.rb (hunk 4 above)
Comment (knorrli) [reply app/controllers/spree/solidus_six_saferpay/transaction/checkout_controller.rb]: 👍

File: app/models/spree/six_saferpay_payment.rb (hunk 5 above)
Comment (knorrli) [reply app/models/spree/six_saferpay_payment.rb:39]: As discussed, this technique of "caching" calculated values in object instances is actually called memoization. Here's an interesting read on the topic: https://thoughtbot.com/blog/ruby-memoization-and-alternatives

File: app/services/spree/solidus_six_saferpay/authorize_payment.rb (hunk 6 above)
Comment (knorrli) [reply app/services/spree/solidus_six_saferpay/authorize_payment.rb]: After our discussion this morning I reviewed the API documentation again and did some testing and I believe you are right. Will remove.

File: app/services/spree/solidus_six_saferpay/initialize_payment.rb (hunk 7 above)
Comment (knorrli) [reply app/services/spree/solidus_six_saferpay/initialize_payment.rb]: removed, see above.

File: lib/solidus_six_saferpay/gateway.rb (hunk 9 above)
Comment (knorrli) [reply lib/solidus_six_saferpay/gateway.rb:62]: see argument here: https://github.com/fadendaten/solidus_six_saferpay/pull/16#discussion_r699993804

File: lib/solidus_six_saferpay/payment_initialize_params.rb (hunk 11 above)
Comment (knorrli) [reply lib/solidus_six_saferpay/payment_initialize_params.rb:58]: Unfortunately this spelling depends on the interface exposed by the [six_saferpay gem](https://github.com/fadendaten/six_saferpay/blob/master/lib/six_saferpay/models/address.rb#L31).
I have opened an issue here: https://github.com/fadendaten/six_saferpay/issues/23

File: spec/services/spree/solidus_six_saferpay/cancel_authorized_payment_spec.rb (hunk 13 above)
Comment (knorrli) [reply spec/services/spree/solidus_six_saferpay/cancel_authorized_payment_spec.rb:31]: This line creates a "test spy" that stubs the `:void` method on this object. This is necessary so that we can later test if the method has been called with `expect(gateway).to have_received(:void)`.

This "spy-and-stub" style of writing tests is the preferred way of writing tests for solidus extensions according to their rubocop configuration. You will find a more in-depth explanation here: https://thoughtbot.com/blog/a-closer-look-at-test-spies

File: spec/models/spree/six_saferpay_payment_spec.rb (hunk 12 above)
Comment (knorrli) [reply spec/models/spree/six_saferpay_payment_spec.rb]: changed spec description to be more descriptive

File: lib/solidus_six_saferpay/address_name_extractor.rb (hunk 8 above)
Comment (knorrli) [reply lib/solidus_six_saferpay/address_name_extractor.rb]: IMO the naming is not wrong since it does in fact extract multiple names (first- and lastname), but I totally agree that it is confusing. I changed the name to `extract_first_and_last_name`.

File: app/controllers/spree/solidus_six_saferpay/checkout_controller.rb (hunk 2 above)
Comment (knorrli) [reply app/controllers/spree/solidus_six_saferpay/checkout_controller.rb]: Good catch! I extracted this into `SixSaferpayPayment.current_payment(@order)`

**Comment by knorrli** [comment]
@Drinion thanks for the review, I fixed/commented on all questions or issues.
